### PR TITLE
Process wide redcap data to long

### LINKED ISSUE #10: happy path preprocessing and merging of two years of redcap data
preprocessing of input file

- convert fields that end with `_timestamp` to dates
- create `rrcp_region` as the non-NA value from `rccp_area_eng`,`rccp_area_ire`, `rccp_area_mal` and `rccp_area_sco` (or just all values whcih start with `rccp_area_`
- create school ID column for non-null value of columans from `entry_school_eng_bark_v1` to `entry_school_sco_wlot_v1`. In the example script a column for each pupil within the school has been created (`v1` to `v16`, but I think just using the first pupil's school seems reasonable?
- [x] @stefpiatek to check that we can use the first school details
- filter out rows where `no_rr_children` is NA or 0

population of output data (BO file)

- convert from wide to long, there as a maximum of 16 pupils for a teacher. I'll note down the steps already taken, but may be an easier way to tranform data. Imagine `pandas.wide_to_long` or `pandas.melt` should be able to do what we want, but I haven't given much thought
- teacher and pupil numbers
  - added `pupilteacherno` by pasting`f"{rccp_rr_id}_{pupil}"` for a range of pupils in `no_rr_children`
  - added `recordno` as the `rccp_rr_id` (they selected this from the `pupilteacherno` in the output data)
  - added `pno` as the `pupil` number, they stripped this out from `pupilteacherno`
- populate correct `reg_rr_title`, `reg_year_rr_trained`, `reg_rr_teacher_status`, `rrcp_country`, 
- populate`rrcp_area` from `rrcp_region`
- For each of the remaining columns in wide data (not including columns already populated and `3_month_followup_timestamp`, `6_month_followup_timestamp`, `national_assessment_i_timestamp`, `national_assessment_ii_timestamp`, national_assessment_iii_timestamp`, `interrupted` and `total_missed_lessons`)
  - taking this from `f"{output_column}_v{pno}"`
- determine total lessons missed by adding all columns starting with `exit_lessons_missed_` (e.g. `exit_lessons_missed_ca`, `exit_lessons_missed_cu`...)
- read in that years school list and convert from code into a label. Example of file and I'll ask for that to be sent to us
    ```
    RRED School ID,School Name
    RRS1010010,Abbots Hall Primary School
    ```
- remove any rows where `rrcp_school` is NA

### PR DESCRIPTION
**Pull request by stefpiatek**
Need to implement conversion of factors (codes or numbers) into strings, but wanted to get some feedback on its current state first. Mostly fine, though a lot of hardcoding columns, but I think that's largely fine as unit tests cover it, so assuming we change test files if the input changes, then we're covered 🤞 

Edited some of the masterfile defintions so that the order matched easily with the expected masterfile, and added in some functionality to help find what the output columns should be

closes  #10

### CODE REVIEW

File: src/rred_reports/masterfile.py [hunk 1]
```diff
@@ -16,7 +16,7 @@ class Pupil(AsFrame):
 
     pupil_no: Data[str]
     rred_user_id: Data[str]
-    assessi_engtest2: Data[pd.Int32Dtype]
+    assessi_engtest2: Data[pd.Int32Dtype]  # surprising that there isn't an engtest1 here, is this an accident?
```
Comment (stefpiatek) [inline src/rred_reports/masterfile.py:19]: I've thought about this more and I guess its not. but still strange

File: tests/test_redcap.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    When the extract is processed
+    Then data coerced from multiple columns and data without students or RR id should be filtered out
+    """
+    file_path = data_path / "redcap" / "extract.csv"
+    extract = pd.read_csv(file_path)
+
+    redcap = preprocess_wide_data(extract)
+
+    # coerced variables filled
+    same_coerced_values = redcap.loc[redcap["record_id"].isin(["AB9234", "AB9235", "AB9236"])]
```
Comment (stefpiatek) [inline tests/test_redcap.py:20]: oh I should also have these as having the same `record_id`

File: src/rred_reports/masterfile.py [hunk 3]
```diff
@@ -142,5 +153,13 @@ def join_masterfile_dfs(masterfile_dfs: dict[str, pd.DataFrame]) -> pd.DataFrame
...
-    pupil_teachers = pd.merge(masterfile_dfs["pupils"], masterfile_dfs["teachers"], on="rred_user_id")
-    return pd.merge(pupil_teachers, masterfile_dfs["schools"], on="school_id")
+    teacher_schools = pd.merge(masterfile_dfs["teachers"], masterfile_dfs["schools"], on="school_id")
+    return pd.merge(teacher_schools, masterfile_dfs["pupils"], on="rred_user_id")
+
+
+def masterfile_columns():
+    pupil_no, user_id, *other_pupil_fields = Pupil.fields()
+    school_id, *other_school_fields = School.fields()
+    user_id, *other_teacher_fields, school_id = Teacher.fields()
```
Comment (harryjmoss) [inline src/rred_reports/masterfile.py:163]: do we want to return `school_id` too?

File: src/rred_reports/masterfile.py [hunk 4]
```diff
@@ -89,15 +92,23 @@ class Teacher(AsFrame):
...
 
     school_id: Data[str]
-    rrcp_school: Data[str]
-    rrcp_area: Data[int]
     rrcp_country: Data[int]
+    rrcp_area: Data[int]
+    rrcp_school: Data[str]
+
+    @classmethod
+    def fields(cls):
```
[the remark this answers is not in the document]
Comment (stefpiatek) [reply src/rred_reports/masterfile.py:110]: 👍 Yeah that'd be nicer for sure

File: src/rred_reports/masterfile.py (hunk 3 above)
Comment (stefpiatek) [reply src/rred_reports/masterfile.py:163]: yeah TBF one of the `school_id`s are just ignored, but it was to be specific about what fields they were. could always just use `_` here instead to make it explicit that we don't care about it

File: src/rred_reports/redcap/main.py [hunk 5]
```diff
@@ -1 +1,174 @@
...
+    _fill_region_and_school_with_coersion(processed_extract)
+    filtered = _filter_out_no_children_and_no_rr_id(processed_extract)
+    return _rename_wide_cols_with_student_number_suffix(filtered)
+
+
+def _convert_timestamps_to_dates(extract: pd.DataFrame):
+    timestamp_cols = [col for col in extract if col.endswith("_timestamp")]
+    dates = (
+        extract[timestamp_cols].applymap(pd.to_datetime, format="%Y-%m-%d %H:%M:%S", errors="coerce")
+        # mapping to date causes NaT to be converted to 2001-01-01
```
Comment (harryjmoss) [inline src/rred_reports/redcap/main.py:26]: interesting!

File: src/rred_reports/redcap/main.py [hunk 6]
```diff
@@ -1 +1,174 @@
...
+    for col in extract:
+        if not col.startswith("pupil_"):
+            continue
+        _, student_number, *col_name_parts = col.split("_")
+        pupil_column_rename[col] = "_".join(col_name_parts) + f"_v{student_number}"
+    return extract.rename(pupil_column_rename, axis=1)
+
+
+# Hardcoded columns for exporting, could finesse this but probably isn't worth the time
+# The final columns output are under unit testing so will catch any changes to input or output data
```
Comment (harryjmoss) [inline src/rred_reports/redcap/main.py:55]: 👍

File: src/rred_reports/masterfile.py (hunk 3 above)
Comment (stefpiatek) [reply src/rred_reports/masterfile.py:159]: oh now I see what you meant, yeah definitely did want to return that 🤦

File: src/rred_reports/masterfile.py [hunk 7]
```diff
@@ -142,5 +150,14 @@ def join_masterfile_dfs(masterfile_dfs: dict[str, pd.DataFrame]) -> pd.DataFrame
...
-    return pd.merge(pupil_teachers, masterfile_dfs["schools"], on="school_id")
+    teacher_schools = pd.merge(masterfile_dfs["teachers"], masterfile_dfs["schools"], on="school_id")
+    return pd.merge(teacher_schools, masterfile_dfs["pupils"], on="rred_user_id")
+
+
+def masterfile_columns():
+    """List of all masterfile columns, in the expected order"""
+    pupil_no, user_id, *other_pupil_fields = Pupil.fields()
+    school_id, *other_school_fields = School.fields()
+    user_id, *other_teacher_fields, school_id = Teacher.fields()
```
Comment (harryjmoss) [inline src/rred_reports/masterfile.py:161]: Might be worth prepending an underscore to one of the `school_id`s here just so it's slightly clearer what the variable contains? I realise it's probably the same value though

File: tests/test_redcap.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    When the extract is processed with the corresponding R script for levels
+    Then data coerced from multiple columns and data without students or RR id should be filtered out
+    """
+    raw_data_path = data_path / "redcap" / "extract.csv"
+    labelled_data_path = data_path / "redcap" / "extract_labels.csv"
+    school_list = data_path / "redcap" / "school_list.csv"
+
+    extract_raw = pd.read_csv(raw_data_path)
+    extract_labelled = pd.read_csv(labelled_data_path)
+    recap_reader = RedcapReader(school_list)
```
Comment (harryjmoss) [inline tests/test_redcap.py:34]:
```suggestion
    redcap_reader = RedcapReader(school_list)
```

File: tests/test_redcap.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    assert (same_coerced_values.get("rrcp_area") == "Bristol").all()
+    assert (same_coerced_values.get("school_id") == "RRS180").all()
+
+    # missing date converted to 0001-01-01
+    assert (redcap.loc[redcap["record_id"] == "AB100"].get("_test_timestamp") == pd.Timestamp.date(pd.NaT)).all()
+    # missing values filtered out
+    assert redcap.loc[redcap["record_id"].isin(["AB101", "AB102", "AB103"])].size == 0
+
+
+def test_read_recap_extract_rows_and_cols(redcap_extract):
```
Comment (harryjmoss) [inline tests/test_redcap.py:48]:
```suggestion
def test_read_redcap_extract_rows_and_cols(redcap_extract):
```

**Review by harryjmoss — APPROVED** [review]
Looks good! Just had some very minor comments/ spotted a couple of typos

File: src/rred_reports/masterfile.py (hunk 7 above)
Comment (stefpiatek) [reply src/rred_reports/masterfile.py]: good call, have made them different and asserted that they are the same as a sanity check

**Pull request merged**